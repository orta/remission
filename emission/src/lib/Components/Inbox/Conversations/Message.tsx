import moment from "moment"
import * as React from "react"
import { View } from "react-native"
import Hyperlink from "react-native-hyperlink"
import { createFragmentContainer, graphql } from "react-relay"
import styled from "styled-components/native"

import colors from "../../../../data/colors"
import SwitchBoard from "../../../NativeModules/SwitchBoard"
import DottedLine from "../../DottedLine"
import { BodyText, FromSignatureText, MetadataText, SmallHeadline } from "../Typography"
import Avatar from "./Avatar"
import ImagePreview from "./Preview/Attachment/ImagePreview"
import PDFPreview from "./Preview/Attachment/PDFPreview"
import InvoicePreview from "./Preview/InvoicePreview"

const VerticalLayout = styled.View`
  flex-direction: column;
  flex: 1;
`

const HorizontalLayout = styled.View`flex-direction: row;`

const Container = styled(View)`
  padding-left: 20;
  padding-right: 20;
`

const Content = styled(HorizontalLayout)`
  align-self: stretch;
  margin-top: 15;
  margin-bottom: 10;
`

const Header = styled(HorizontalLayout)`
  align-self: stretch;
  margin-bottom: 10;
`

const TextContainer = styled(VerticalLayout)`
  margin-left: 10;
`

const SenderName = styled(SmallHeadline)`
  marginRight: 6
  font-size: 11.5
`

const FromSignature = styled(FromSignatureText)`
  marginTop: 10
`

interface TimeStampProps {
  pending: boolean
}

const TimeStamp = styled(MetadataText)`
  font-size: 11.5
  color: ${(p: TimeStampProps) => (p.pending ? colors["yellow-bold"] : colors["gray-medium"])}
`

const Seperator = styled(DottedLine)`
  padding-left: 20;
  padding-right: 20;
`

const PreviewContainer = styled.View`margin-bottom: 10;`

interface Props extends RelayProps {
  senderName: string
  initials?: string
  artworkPreview?: JSX.Element
  showPreview?: JSX.Element
  firstMessage: boolean
  index: number
  initialText: string
}

export class Message extends React.Component<Props, any> {
  renderAttachmentPreviews(attachments: Props["message"]["attachments"]) {
    // This function does not use the arrow syntax, because it shouldn’t be force bound to this component. Instead, it
    // gets bound to the AttachmentPreview component instance that’s touched, so we can pass `this` to `findNodeHandle`.
    const previewAttachment = function(attachmentID) {
      const attachment = attachments.find(({ id }) => id === attachmentID)
      SwitchBoard.presentMediaPreviewController(this, attachment.download_url, attachment.content_type, attachment.id)
    }

    return attachments.map(attachment => {
      if (attachment.content_type.startsWith("image")) {
        return (
          <PreviewContainer key={attachment.id}>
            <ImagePreview attachment={attachment as any} onSelected={previewAttachment} />
          </PreviewContainer>
        )
      }
      if (attachment.content_type === "application/pdf") {
        return (
          <PreviewContainer key={attachment.id}>
            <PDFPreview attachment={attachment as any} onSelected={previewAttachment} />
          </PreviewContainer>
        )
      }
    })
  }

  renderBody() {
    const { message, firstMessage, initialText } = this.props
    const isSent = !!message.created_at
    const body = firstMessage ? initialText : message.body

    const onLinkPress = url => {
      return SwitchBoard.presentNavigationViewController(this, url)
    }

    const linkStyle = {
      color: "#0645ad",
    }

    return (
      <Hyperlink onPress={onLinkPress} linkStyle={linkStyle}>
        <BodyText disabled={!isSent}>
          {body}
        </BodyText>
      </Hyperlink>
    )
  }

  render() {
    const { artworkPreview, initials, message, senderName, showPreview } = this.props
    const isPending = !message.created_at

    const fromName = message.from.name
    const fromEmail = message.from.email

    const fromSignature = fromName ? `${fromName} · ${fromEmail}` : fromEmail
    return (
      <Container>
        <Content>
          <Avatar isUser={message.is_from_user} initials={initials} />
          <TextContainer>
            <Header>
              <SenderName disabled={isPending}>
                {senderName}
              </SenderName>
              {
                <TimeStamp pending={isPending}>
                  {isPending ? "pending" : moment(message.created_at).fromNow(true)}
                </TimeStamp>
              }
            </Header>
            {artworkPreview &&
              <PreviewContainer>
                {artworkPreview}
              </PreviewContainer>}

            {showPreview &&
              <PreviewContainer>
                {showPreview}
              </PreviewContainer>}

            {message.invoice &&
              <PreviewContainer>
                <InvoicePreview invoice={message.invoice} />
              </PreviewContainer>}

            {this.renderAttachmentPreviews(message.attachments)}

            {this.renderBody()}

            {!message.is_from_user &&
              <FromSignature>
                {fromSignature}
              </FromSignature>}
          </TextContainer>
        </Content>
        {this.props.index !== 0 && <Seperator />}
      </Container>
    )
  }
}

export default createFragmentContainer(
  Message,
  graphql`
    fragment Message_message on Message {
      body
      created_at
      is_from_user
      from {
        name
        email
      }
      invoice {
        ...InvoicePreview_invoice
      }
      attachments {
        id
        content_type
        download_url
        file_name
        ...ImagePreview_attachment
        ...PDFPreview_attachment
      }
    }
  `
)

interface RelayProps {
  message: {
    body: string | null
    created_at: string | null
    is_from_user: boolean
    invoice: any | null
    from: {
      name: string | null
      email: string
    }
    attachments: Array<{
      id: string
      content_type: string
      download_url: string
    }>
  }
}
