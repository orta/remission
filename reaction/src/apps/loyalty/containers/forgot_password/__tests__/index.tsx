import { shallow } from "enzyme"
import * as React from "react"
import * as renderer from "react-test-renderer"

jest.mock("isomorphic-fetch")
const fetch = require("isomorphic-fetch").default

import Button from "../../../../../components/buttons/inverted"
import Message from "../../../../../components/message"
import ForgotPassword from "../index"

describe("<ForgotPassword />", () => {
  const APP_TOKEN = "force-staging"
  const SUBMIT_URL = "https://stagingapi.artsy.net/api/v1/users/send_reset_password_instructions"

  afterEach(fetch.mockReset)

  it("renders the snapshot", () => {
    const component = renderer.create(<ForgotPassword appToken={APP_TOKEN} submitEmailUrl={SUBMIT_URL} />)
    expect(component).toMatchSnapshot()
  })

  it("displays correct message on success", done => {
    fetch.mockImplementation(() => Promise.resolve({ status: 201 }))

    const expectedResult = (
      <span>
        Instructions on how to reset your password have been sent to <b>test@artsymail.com</b>
      </span>
    )
    const wrapper = shallow(<ForgotPassword appToken={APP_TOKEN} submitEmailUrl={SUBMIT_URL} />)
    wrapper.setState({ email: "test@artsymail.com" })
    const button = wrapper.find(Button)
    button.simulate("click")

    setImmediate(() => {
      const message = wrapper.find(Message)
      expect(message.contains(expectedResult)).toBeTruthy()
      done()
    })
  })

  it("displays correct message on failure", done => {
    fetch.mockImplementation(() => Promise.resolve({ status: 400 }))

    const expectedResult = <span>No account exists for <b>test@artsymail.com</b></span>
    const wrapper = shallow(<ForgotPassword appToken={APP_TOKEN} submitEmailUrl={SUBMIT_URL} />)
    wrapper.setState({ email: "test@artsymail.com" })
    const button = wrapper.find(Button)
    button.simulate("click")

    setImmediate(() => {
      const message = wrapper.find(Message)
      expect(message.contains(expectedResult)).toBeTruthy()
      done()
    })
  })
})
