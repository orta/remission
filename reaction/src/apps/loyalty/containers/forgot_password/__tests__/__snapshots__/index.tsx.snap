// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPassword /> renders the snapshot 1`] = `
<div
  className="file__Container-s1up4l80-0 eEevDd"
>
  <div
    className="file-s1ekx40u-0 iuEcGN"
    style={undefined}
  >
    î˜‘
  </div>
  <p
    className="file__Text-s4rtu63-0 eEhqjQ"
  >
    Enter the email address associated
    <br />
    with your account.
  </p>
  <div
    className="message-box"
  >
    
  </div>
  <input
    autoFocus={true}
    className="file__StyledInput-s1up4l80-1 koOAAW file__StyledInput-s2fklig-0 ffQmtI"
    name="email"
    onChange={[Function]}
    placeholder="Email"
    value=""
  />
  <button
    className="file__InvertedButton-s2qkmvo-0 hmYmGS file__StyledButton-muis26-0 kaBVMO"
    disabled={true}
    onClick={[Function]}
  >
    <span>
      Reset Password
    </span>
  </button>
</div>
`;
